File: examples/inputs/start-workflow/support.py
import collections.abc
import typing

import temporalio.workflow

_ResultT = typing.TypeVar("_ResultT")


class StartedWorkflowHandle(typing.Generic[_ResultT]):
    def __init__(
        self,
        handle: temporalio.workflow.ExternalWorkflowHandle[_ResultT],
    ) -> None:
        self._handle: temporalio.workflow.ExternalWorkflowHandle[_ResultT] = handle

    @property
    def workflow_id(self) -> str:
        return self._handle.id

    @property
    def run_id(self) -> str | None:
        return self._handle.run_id

    async def cancel(self) -> None:
        await self._handle.cancel()

    async def signal(
        self,
        signal: str | collections.abc.Callable[..., typing.Any],
        *args: typing.Any,
    ) -> None:
        await self._handle.signal(signal, *args)

    async def get_result(self) -> _ResultT:
        raise NotImplementedError(
            "result retrieval is not yet implemented for started workflow handles"
        )


def started_workflow_handle_from_proto(
    workflow_id: str,
    run_id: str | None,
) -> StartedWorkflowHandle[typing.Any]:
    handle = temporalio.workflow.get_external_workflow_handle(
        workflow_id,
        run_id=run_id,
    )
    return StartedWorkflowHandle(handle)

File: examples/inputs/start-workflow/support.ts
export class StartedWorkflowHandle {
  public readonly workflowId: string;
  public readonly runId: string | undefined;

  public constructor(
    private readonly handle: workflow.ExternalWorkflowHandle,
  ) {
    this.workflowId = handle.workflowId;
    this.runId = handle.runId;
  }

  public async cancel(): Promise<void> {
    await this.handle.cancel();
  }

  public async signal<Args extends any[] = [], Name extends string = string>(
    def: workflow.SignalDefinition<Args, Name> | string,
    ...args: Args
  ): Promise<void> {
    await this.handle.signal(def, ...args);
  }

  public async getResult(): Promise<unknown> {
    throw new Error(
      "result retrieval is not yet implemented for started workflow handles",
    );
  }
}

export function startedWorkflowHandleFromProto(
  workflowId: string,
  runId: string | undefined,
): StartedWorkflowHandle {
  return new StartedWorkflowHandle(
    workflow.getExternalWorkflowHandle(workflowId, runId),
  );
}
